<a href="https://colab.research.google.com/github/lumagri/Data_Science/blob/main/Python_Data_Science.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python para Data Science

Notebook para o curso de Python para Data Science publicado na plataforma da [Alura](https://www.alura.com.br/)

# Introdução ao Python

Vamos conhecer o Python, nosso ambiente de estudo e também faremos nosso primeiro código nessa linguagem!

## Google Colaboratory

Já aprendemos o que é o [Python](https://www.python.org/) agora vamos aprender a utilzar o [Google Colaboratory](https://colab.research.google.com/) e programar em um notebook.

Vamos testar algumas funções dessa ferramenta.

 **`>>> Use essa célula para mover <<<`**

In [ ]:
10

10

10

## Olá mundo!

Vamso conhecer mais como funciona o ambiente interativo de um notebook. Para isso vamos executar nosso primeiro comando em Python: [`print()`](https://docs.python.org/3/library/functions.html#print)

In [ ]:
print('Olá Mundo')

Olá Mundo


In [ ]:
print(10)

10


In [ ]:
print('Luciane', 37)

Luciane 37


In [ ]:
print('Escola de dados Alura')

Escola de dados Alura


In [ ]:
print('Nome: [Luciane]'),
print('Sobrenome: [Magri Tomaz]')

Nome: [Luciane]
Sobrenome: [Magri Tomaz]


In [ ]:
print('L')
print('u')
print('c')
print('i')
print('a')
print('n')
print('e')


L
u
c
i
a
n
e


In [ ]:
print(25, 'Fevereiro', 1987)

25 Fevereiro 1987


In [ ]:
print('Ano atual:', 2024)

Ano atual: 2024


# Configurações de Texto #

**Texto em negrito**
*Texto em Itálico*
***Texto em negrito e itálico***

##Listas não numeradas##
* Item 1
* Item da lista
* Item da lista

##Listas numeradas##
1. item 1
2. item 2

## Para transformar um conteúdo em link ##
Colocar entre parênteses (https://www.alura.com.br/)




# Manipulando dados

Vamos aprender sobre as variáveis no python, como elas são declaradas e utilizadas além de conhecer outros comandos dentro do Python. :D

## Variáveis

Em Data Science nós trabalhamos com vários dados e informações, então é essencial saber trabalhar com variáveis.

Criamos uma variável no python através da atribuição de um valor a ela.

Para fazer isso, colocamos o nome da variável um sinal de igual (`=`) e o valor que queremos atribuir

In [ ]:
idade = 5

In [ ]:
print(idade)

5


In [ ]:
idade = 10
print(idade)

10


In [ ]:
idade = 15
idade

15

In [ ]:
nome = 'Gabriel'
nome

'Gabriel'

Nomes que **não** podemos definir para variáveis:

- **Nomes que começam com números**
  - Exemplos: `10_notas`, `2_nomes_casa`, etc.
- **Palavras separada por espaço**
  - Exemplos: `Nome escola`, `notas estudantes`, etc.
- **Nomes de funções do Python**
  - Exemplos: `print`, `type`, etc.

> Letras maiúsculas e minúsculas vão gerar diferentes variáveis. A varíavel `idade` é diferente de `Idade` que por sua vez também é diferente de `IDADE`:
``` Python
idade = 1
Idade = 2
IDADE = 3
_idade = 4
_idade_ = 5
print(idade, Idade, IDADE, _idade, _idade_)
1 2 3 4 5
```

## Tipos de variáveis

Cada variável contém uma classe especifica quanto ao tipo de objeto que ela está se referenciando. Essas classes vão ser diferentes a partir do tipo de dado que nós atribuimos a uma variável.

Para saber a classe de cada elemento usamos a função [`type()`](https://docs.python.org/3/library/functions.html#type)

In [ ]:
i = 5
type(i)

int

In [ ]:
f = 9.8
type(f)

float

In [ ]:
s = 'Luciane'
type(s)

str

In [ ]:
b = True
type(b)

bool

Em um conjunto de dados escolares podemos ter vários tipos de informações. Digamos que tenhamos acesso à ficha de dados do aluno *Frabicio Daniel* como transformamos ela em variáveis no Python?

#### Ficha:

- Nome: Fabricio Daniel
- Idade: 15 anos
- Media do semestre: 8,45
- Situação de aprovação: Verdadeira (aprovado)

In [ ]:
nome_aluno = 'Fabricio Daniel'
idade_aluno = 15
media_aluno = 8.45
situacao_aprovado = True
print(nome_aluno,idade_aluno,media_aluno,situacao_aprovado)

Fabricio Daniel 15 8.45 True


## Variáveis numéricas

Entre os tipos de dados numéricos vamos nos focar no tipo `inteiro` e `float`.

Temos uma tabela de informação de empregos quanto ao cargo, quantidade de pessoas empregadas e o salário correspondente:

|Cargo | Quantidade | Salário|
|---|---|---|
|Segurança | 5 | 3000 |
|Docente | 16| 6000|
|Diretoria| 1 |12500|

Precisamos trabalhar com esses dados fornecendo:

- A quantidade total de empregados;
- A diferença entre o salário mais baixo e mais alto; e
- A média ponderada da faixa salarial da escola.

In [ ]:
q_seguranca = 5
s_seguranca = 3000

q_docente = 16
s_docente = 6000

q_diretoria = 1
s_diretoria = 12500

In [ ]:
total_empregados = q_seguranca + q_docente + q_diretoria
total_empregados

22

In [ ]:
diferenca_salario =  s_diretoria - s_seguranca
diferenca_salario

9500

In [ ]:
media = (q_seguranca*s_seguranca + q_docente*s_docente + q_diretoria*s_diretoria) / (total_empregados)
media

5613.636363636364

## Strings

Strings são caracterizada por ter um conjunto de caracteres formando um texto.

Strigs podem ser criadas ao atribuirmos a uma variável um dado que esteja entre aspas simples (`'`) ou aspas duplas (`"`)

In [ ]:
s1 = 'Alura'
s2 = "Alura"
print(type(s1),type(s2))

<class 'str'> <class 'str'>


As variáveis textuais contém vários métodos que nos ajudam a formatar strings. Métódos podem ser executados ao definirmos um objeto seguindo a seguinte estrutura:

```
objeto.metodo()
```

Existem métodos que não necessitam dos `()`, é preciso verificar a documentação de cada caso.

---
**Situação:**

Recebemos uma variável com o nome de uma professora da escola para inserimos no cadastro. No entanto, precisamos tratar esse texto antes de inserirmos no sistema

In [ ]:
texto = '  Geovana Alessandra dias Sanyos '

O objetivo final é que o nome esteja da seguinte forma:

```
'GEOVANA ALESSANDRA DIAS SANTOS'
```

### [`str.upper()` ](https://docs.python.org/3/library/stdtypes.html#str.upper)
Converte uma string para maiúsculas

In [ ]:
texto.upper()

'  GEOVANA ALESSANDRA DIAS SANYOS '

### [`str.lower()`](https://docs.python.org/3/library/stdtypes.html#str.lower)
Método converte uma string para minúsculas.

In [ ]:
texto.lower()

'  geovana alessandra dias sanyos '

### [`str.strip()`](https://docs.python.org/3/library/stdtypes.html#str.strip)
Método remove os espaços em branco do início e do fim de uma string.

In [ ]:
texto.strip()

'Geovana Alessandra dias Sanyos'

### [`str.replace(antigo, novo)`](https://docs.python.org/3/library/stdtypes.html#str.replace)

Método substitui todas as ocorrências do texto "antigo" na string por "novo"

In [ ]:
texto.replace('y','t')

'  Geovana Alessandra dias Santos '

### Observações

1. Os métodos retornam uma **tranformação**, não a executam no texto!

2. Além disso, podemos acumular a execução de métodos.

In [ ]:
texto

'  Geovana Alessandra dias Sanyos '

Para que seja executada a transformação nós podemos atribuir às saídas das transformações à variável

In [ ]:
texto = texto.strip().replace('y','t').upper()
texto

'GEOVANA ALESSANDRA DIAS SANTOS'

## Coletando dados

Em algumas aplicações precisamos coletar valores da pessoa usuária do nosso projeto. Em python conseguimos coletar dados de usuário através do comando [`input()`](https://docs.python.org/3/library/functions.html#input).

Para fazer essa coleta podemos atribuir essa função à uma variável.

In [ ]:
nome = input('Escreva seu nome: ')

Escreva seu nome: Luciane Magri


In [ ]:
nome

'Luciane Magri'

O retorno desse comando sempre será uma *string*. Isso quer dizer que mesmo que façamos uma coleta de algo que deva ser numérico, ele será uma string.

Então, será preciso **converter o resultado caso não seja desejável obter uma string**.

Existem funções para conversão de valores:

- Inteiros: [`int(dado_para_conversao)`](https://docs.python.org/3/library/functions.html#int)
- Float: [`float(dado_para_conversao)`](https://docs.python.org/3/library/functions.html#float)
- String: [`str(dado_para_conversao)`](https://docs.python.org/3/library/functions.html#func-str)
- Booleano: [`bool(dado_para_conversao)`](https://docs.python.org/3/library/functions.html#bool)

In [ ]:
ano_entrada = input('Escreva o ano de ingresso do(a) estudante: ')

Escreva o ano de ingresso do(a) estudante: 2024


In [ ]:
type(ano_entrada)

str

In [ ]:
ano_entrada = int(input('Escreva o ano de ingresso do(a) estudante: '))

Escreva o ano de ingresso do(a) estudante: 2024


In [ ]:
type(ano_entrada)

int

Buscaremos apresentar melhor agora o resultado que obtivemos da transformação. Nós conseguimos **formatar** e apresentar o nosso resultado misturando strings com valores não textuais.

Para fazer isso usamos a estrutura de formatação `f` com strings.


In [ ]:
nota_entrada = float(input('Digite a nota do teste de ingresso: '))
print(f'Ano de entrada {ano_entrada} - nota do teste de ingresso {nota_entrada}')

Digite a nota do teste de ingresso: 9.0
Ano de entrada 2024 - nota do teste de ingresso 9.0


Durante a aula, aprendemos a utilizar a formatação f-string (ou formatação de string), na qual colocamos um f antes da criação da string e as variáveis entre chaves {}.

Mas existem outras formatações, como o uso do operador de formatação de string ou da função .format(). Esse operador de formatação permite a inserção de variáveis em pontos específicos na string com o operador %. Esse operador funciona como um marcador, informando onde o valor da variável vai ser exposto na string.

O % precisa ser acompanhado de uma palavra-chave para cada tipo de variável que se deseja adicionar.

In [ ]:
nome_aluno = 'Fabricio Daniel'

print('Nome do aluno: %s' %(nome_aluno))


Nome do aluno: Fabricio Daniel


In [ ]:
nome_aluno = 'Fabricio Daniel'
idade_aluno = 15
media_aluno = 8.45

print('Nome do aluno é %s, ele tem %d anos e sua média é %f.' %(nome_aluno, idade_aluno, media_aluno))


Nome do aluno é Fabricio Daniel, ele tem 15 anos e sua média é 8.450000.


# Execícios



1.   Crie um programa que solicite à pessoa usuária digitar seu nome, e imprima “Olá, [nome]!”.

2.   Crie um programa que solicite à pessoa usuária digitar seu nome e idade, e imprima “Olá, [nome], você tem [idade] anos.”.

3.  Crie um programa que solicite à pessoa usuária digitar seu nome, idade e altura em metros, e imprima “Olá, [nome], você tem [idade] anos e mede [altura] metros!”.



In [ ]:
nome = input("Qual o seu nome? ")
print(f'Olá {nome}')

Qual o seu nome? Luciane
Olá Luciane


In [ ]:
nome = input ("Qual seu nome?")
idade = input("Qual sua idade?")
print(f'Olá {nome}, você tem {idade}.')

Qual seu nome?Luciane 
Qual sua idade?37
Olá Luciane , você tem 37.


In [ ]:
type(idade)

str

In [ ]:
nome = input ("Qual seu nome? ")
idade = input("Qual sua idade? "); idade = int(idade)
altura = input("Qual sua altura? "); altura = float(altura);

print( f'Olá %s, você tem %i anos e mede %f metros.' %(nome, idade, altura))

Qual seu nome?Luciane
Qual sua idade?37
Qual sua altura?1.78
Olá Luciane, você tem 37 anos e mede 1.780000 metros.


In [ ]:
print(type(idade),type(nome),type(altura))

<class 'int'> <class 'str'> <class 'float'>


In [ ]:
print( f'Olá %s, você tem %s anos e mede %s metros.' %(nome, idade, altura))

Olá Luciane, você tem 37 anos e mede 1.78 metros.


# Exercício 2



## Calculadora com operadores ##

1. Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a soma dos dois valores.

In [ ]:
num1 = input("Digite 1 número " ); num1 = int(num1);
num2 = input("Digite outro número " ); num2 = int(num2);

print(f'A soma do primeiro número %i  +  o segundo número %i  é igual a ' %(num1, num2), num1+num2);


Digite 1 número 10
Digite outro número 0
A soma do primeiro número 10  +  o segundo número 0  é igual a  10


2. Crie um programa que solicite três valores numéricos à pessoa usuária e imprima a soma dos três valores.

In [ ]:
num1 = input("Digite 1 número " ); num1 = int(num1);
num2 = input("Digite outro número " ); num2 = int(num2);
num3 = input("Digite o terceiro número " ); num3 = int(num3);

print(f'A soma dos 3 números %i  + %i + %i  é igual a ' %(num1, num2, num3), num1+num2+num3);


Digite 1 número 10
Digite outro número 10
Digite o terceiro número 10
A soma dos 3 números 10  + 10 + 10  é igual a  30


3. Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a subtração do primeiro pelo o segundo valor.


In [ ]:
nx = input("Digite 1 número " ); nx = int(nx);
ny = input("Digite outro número " ); ny = int(ny);
print(f'A subtração de %i por %i é igual a => '%(nx,ny), nx-ny);

Digite 1 número 20
Digite outro número 30
A subtração de 20 por 30 é igual a =>  -10



4. Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a multiplicação dos dois valores.



In [ ]:
nx = input("Digite 1 número " ); nx = int(nx);
ny = input("Digite outro número " ); ny = int(ny);
print(f'A multiplicação de %i por %i é = a '%(nx,ny), nx*ny)

Digite 1 número 5
Digite outro número 10
A multiplicação de 5 por 10 é = a  50


5. Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e realize a divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser 0.


In [ ]:
nx = input("Digite 1 número " ); nx = float(nx);
ny = input("Digite outro número diferente de 0 para ser dividido " ); ny = float(ny);
print(f'A divisão de %.2f. por %2f. é =  '%(nx,ny),nx/ny)

Digite 1 número 10
Digite outro número diferente de 0 para ser dividido 3
A divisão de 10.00. por 3.000000. é =   3.3333333333333335



6. Crie um programa que solicite dois valores numéricos, um operador e uma potência, e realize a exponenciação entre esses dois valores.


In [ ]:
nx = input("Digite 1 número " ); nx = int(nx);
ny = input("Digite outro número " ); ny = int(ny);
print(f'O valor de %i elevado a %i é = '%(nx,ny), nx**ny)

Digite outro número 2
O valor de 10 elevado a 2 é =  100



7. Crie um programa que solicite dois valores numéricos, um numerador e um denominador e realize a divisão inteira entre os dois valores. Deixe claro que o valor do denominador não pode ser 0.


In [ ]:
nx = input("Digite 1 número : " ); nx = int(nx);
ny = input("Digite outro número diferente de 0 para ser dividido : " ); ny = int(ny);
print(f'A divisão de %s por %s é =  '%(nx,ny), nx/ny)

Digite 1 número : 25
Digite outro número diferente de 0 para ser dividido : 2
A divisão de 25 por 2 é =   12.5



8. Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e retorne o resto da divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser 0.



9. Crie um código que solicita 3 notas de um estudante e imprima a média das notas.



In [ ]:
nx = input("Insira a sua  primeira nota : " ); nx = int(nx);
ny = input("Insira a sua  segunda nota: " ); ny = int(ny);
nz = input("Insira a sua  terceira nota: " ); nz = int(nz);
print('A média das suas 3 notas é = ' , ((nx + ny + nz)/3))

Insira a sua  primeira nota : 5
Insira a sua  segunda nota: 5
Insira a sua  terceira nota: 5
A média das suas 3 notas é =  5.0


In [ ]:
type(nz)

int

10. Crie um código que calcule e imprima a média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4.

In [ ]:
n1 = 5; p1= 1;
n2 = 12; p2= 2;
n3 = 20; p3= 3;
n4 = 15; p4= 4;

media_ponderada = ( n1*p1 + n2*p2 + n3*p3 + n4*p4)/ (p1 + p2 + p3 + p4);
print("O valor da média ponderada é : ", media_ponderada)


O valor da média ponderada é :  14.9


# Estruturas condicionais

## `IF` e `ELSE`

O `if` e `else` são duas estruturas condicionais. O `if` executará o bloco de comando caso a condição colocada for **verdadeira**. O `else` é um caso em que a condicional de `if` seja **falsa**.

O `if` é uma palavra-chave em Python que significa "se". Ele é usado para formar uma estrutura condicional, que permite que você verifique se uma determinada condição é verdadeira ou falsa e, em seguida, execute um bloco de código específico dependendo do resultado da verificação. A sintaxe para usar o `if` é:



```
if condição:
    # faça algo
```



In [ ]:
if 2<7:
  print('condição verdadeira')
  print()
print('fora do bloco')

condição verdadeira

fora do bloco


Já o `else` em Python é usada em conjunto com a palavra-chave `if` para formar uma estrutura condicional. A sintaxe para usar o `else` é:

```
if condição:
  # código caso seja verdade
else:
  # código caso seja falso
```

O `else` é executado quando a condição verificada pelo `if` é avaliada como `False`.

In [ ]:
idade_maria = int(input('Digite a idade da Maria: '))
idade_beatriz = int(input('Digite a idade da Beatriz: '))

if idade_maria > idade_beatriz:
  print('Maria é mais velha que Beatriz.')


Digite a idade da Maria: 20
Digite a idade da Beatriz: 30


---
**Situação:**

Receberemos a média de estudantes e precisamos de um algoritmo que execute a análise e decida se esse estudante está **Aprovado** ou Reprovado, mostrando uma mensagem do resultado. Para ser aprovado, a média precisa ser igual ou superior à 6.0.

In [ ]:
media = float(input('Digite a média: '))

if media >= 6.0:
  print('Aprovado(a)')
else:
  print('Reprovado(a)')

Digite a média: 4.0
Reprovado(a)


Agora a nossa instituição de ensino lançou uma nota oficial que pessoas que tenham média entre 4.0 e 6.0 podem fazer os cursos de **Recuperação** nas férias para poder recuperar a nota.

Então podemos agora fazer um conjunto de `if`s para que possamos estruturar essa nova condição.

In [ ]:
media = float(input('Digite a média: '))

if media >= 6.0:
  print('Aprovado(a)')
if 6.0 > media >= 4.0:
  print('Recuperação')
if media < 4.0:
  print('Reprovado(a)')

Digite a média: 3
Reprovado(a)


Notemos que em casos com 3 situações como esse precisamos definir bem nossas condições. Pois foi feita uma construção com `else` no final, ele irá considerar apenas a alguma condicional para ser o caso **falso** podendo resultar em duas (ou mais) execuções.

Por exemplo:

In [ ]:
media = float(input('Digite a média: '))

if media >= 6.0:
  print('Aprovado(a)')
if 6.0 > media >= 4.0:
  print('Recuperação')
else:
  print('Reprovado(a)')

Digite a média: 7.0
Aprovado(a)
Reprovado(a)


## `ELIF`

O `elif` é uma palavra-chave em Python que significa "senão, se" e pode ser considerado uma união do `else` com um `if`. Ela é usada em conjunto com a palavra-chave `if` para formar uma estrutura condicional encadeada.



A sintaxe para usar o `elif` é:

```
if condição1:
    # faça algo
elif condição2:
    # faça outra coisa
elif condição3:
    # faça mais alguma coisa
else:
    # faça algo diferente
```

O `elif` permite que você verifique várias condições de forma encadeada, economizando espaço em seu código. Se a primeira condição for avaliada como `False`, o interpretador Python avaliará a próxima condição no `elif`. Isso continuará até que uma condição seja avaliada como `True` ou até que o `else` seja atingido. Se nenhuma das condições forem avaliadas como `True`, a execução do código do `else` será iniciada.

Vamos usar o mesmo caso anterior:

In [ ]:
media = float(input('Digite a média: '))

if media >= 6.0:
  print('Aprovado(a)')
elif 6.0 > media >= 4.0:
  print('Recuperação')
else:
  print('Reprovado(a)')

Digite a média: 5.0
Recuperação


## Operadores

Durante a construção de comandos por vezes precisamos de uma elaboração maior de da expressão condicional, necessitando que alguns operadores lógicos estejam integrados.

### `AND`, `OR`, `NOT`

Os operadores lógicos `and`, `or` e `not` são usados para combinar expressões lógicas em Python. Eles são usados frequentemente em conjunto com o `if` para criar estruturas condicionais mais complexas.


- `AND` é usado para verificar se duas condições são verdadeiras. A expressão lógica¹ `x and y` é avaliada como `True` apenas se **ambas as condições `x` e `y` forem verdadeiras**, e como `False` caso contrário.

- `OR` é usado para verificar se pelo menos uma das condições é verdadeira. A expressão lógica `x or y` é avaliada como `True` **se pelo menos uma das condições `x` ou `y` for verdadeira**, e como `False` se ambas forem falsas.

- `NOT` é usado para **negar uma condição**. A expressão lógica not x é avaliada como True se a condição x for falsa, e como False se a condição x for verdadeira.

¹ Uma expressão lógica é uma declaração que pode ser avaliada como verdadeira ou falsa. Ela é composta por operandos lógicos² e operadores lógicos³, que são usados ​​para combinar várias expressões lógicas em uma única expressão.

² Os operandos lógicos são os elementos que são comparados ou avaliados em uma expressão lógica. Eles são geralmente valores verdadeiros ou falsos, mas também podem ser expressões lógicas mais complexas. Em Python, os operandos lógicos são os valores `True` e `False`.

³ Os operadores lógicos são os símbolos ou palavras-chave que são usados ​​para combinar várias expressões lógicas em uma única expressão. Em Python, os operadores lógicos são `and`, `or` e `not`, bem como as palavras-chave `if`, `elif` e `else`.

In [ ]:
t1 = t2 = True
f1 = f2 = False

In [ ]:
if t1 and f2:
  print('expressão verdadeira')
else:
  print('expressão falsa')

expressão falsa


In [ ]:
if t1 or f2:
  print('expressão verdadeira')
else:
  print('expressão falsa')

expressão verdadeira


In [ ]:
if not f1:
  print('expressão verdadeira')
else:
  print('expressão falsa')

expressão verdadeira


### `IN`

É usado para verificar se um elemento está presente em uma lista, tupla ou outra variável de conjunto. A expressão `x in y` é avaliada como `True` se o elemento `x` estiver presente na variável de conjunto `y`, e como `False` caso contrário.

Podemos verificar isso com variáveis de texto.

---

**Situação:**

Na escola foi passada uma lista com nomes de estudantes que foram aprovados por média no semestre, mas é preciso verificar se alguns nomes estão nessa lista para verificar se os dados estão corretos.

A lista distribuida pode ser observada abaixo:

```
lista = 'José da Silva, Maria Oliveira, Pedro Martins, Ana Souza, Carlos Rodrigues, Juliana Santos, Bruno Gomes, Beatriz Costa, Felipe Almeida, Mariana Fernandes, João Pinto, Luísa Nascimento, Gabriel Souza, Manuela Santos, Thiago Oliveira, Sofia Ferreira, Rafael Albuquerque, Isabella Gomes, Bruno Costa, Maria Martins, Rafaela Souza, Matheus Fernandes, Luísa Almeida, Beatriz Pinto, Mariana Rodrigues, Gabriel Nascimento, João Ferreira, Maria Albuquerque, Felipe Oliveira
'
```

Os nomes que precisam ser verificados são os seguintes:

```
nome_1 = 'Mariana Rodrigues'
nome_2 = 'Marcelo Nogueira'
```

In [ ]:
lista = 'José da Silva, Maria Oliveira, Pedro Martins, Ana Souza, Carlos Rodrigues, Juliana Santos, Bruno Gomes, Beatriz Costa, Felipe Almeida, Mariana Fernandes, João Pinto, Luísa Nascimento, Gabriel Souza, Manuela Santos, Thiago Oliveira, Sofia Ferreira, Rafael Albuquerque, Isabella Gomes, Bruno Costa, Maria Martins, Rafaela Souza, Matheus Fernandes, Luísa Almeida, Beatriz Pinto, Mariana Rodrigues, Gabriel Nascimento, João Ferreira, Maria Albuquerque, Felipe Oliveira'
lista

'José da Silva, Maria Oliveira, Pedro Martins, Ana Souza, Carlos Rodrigues, Juliana Santos, Bruno Gomes, Beatriz Costa, Felipe Almeida, Mariana Fernandes, João Pinto, Luísa Nascimento, Gabriel Souza, Manuela Santos, Thiago Oliveira, Sofia Ferreira, Rafael Albuquerque, Isabella Gomes, Bruno Costa, Maria Martins, Rafaela Souza, Matheus Fernandes, Luísa Almeida, Beatriz Pinto, Mariana Rodrigues, Gabriel Nascimento, João Ferreira, Maria Albuquerque, Felipe Oliveira'

In [ ]:
nome_1 = 'Mariana Rodrigues'
nome_2 = 'Marcelo Nogueira'

In [ ]:
if nome_1 in lista:
  print(f'{nome_1} está na lista')
else:
  print(f'{nome_1} não está na lista')

Mariana Rodrigues está na lista


In [ ]:
if nome_2 in lista:
  print(f'{nome_2} está na lista')
else:
  print(f'{nome_2} não está na lista')

Marcelo Nogueira não está na lista


# Exercícios : Desafios para praticar



 1) Escreva um programa que peça à pessoa usuária para fornecer dois números e exibir o número maior.

In [ ]:
n1 = input( 'Digite 1 numero: '),
n2 = input( 'Digite outro numero: '),

if n1 > n2:
  print(n1),
else: print(n2);

Digite 1 numero: 10
Digite outro numero: 15
('15',)


2) Escreva um programa que solicite o percentual de crescimento de produção de uma empresa e informe se houve um crescimento (porcentagem positiva) ou decrescimento (porcentagem negativa).

In [ ]:
Percentual_de_crescimento = float(input( 'Digite o percentual de crescimento '))
if Percentual_de_crescimento > 1:
  print(f'O crescimento de {Percentual_de_crescimento:.2f}% é positivo para a empresa')
else:
  print(f'O crescimento de {Percentual_de_crescimento:.2f}% é negativo para a empresa')

Digite o percentual de crescimento -5
O crescimento de -5.00% é negativo para a empresa


3) Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante.

In [ ]:
Letra = input( 'Digite uma letra: ').lower()
if Letra in ['a', 'e' , 'i' , 'o', 'u']:
  print(f'Esta letra {Letra} ' 'é uma vogal')
else:
  print(f'Esta letra {Letra}' ' é uma consoante')


Digite uma letra: o
Esta letra o é uma vogal


4) Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos.

In [ ]:

Ano_do_carro = int(input('Qual o ano do carro? '))
preço_do_carro_ano1 = int(input(f'Qual o preço do carro em {Ano_do_carro} '))
Ano_do_carro2 = int(Ano_do_carro + 1)
preço_do_carro_ano2 = int(input(f'Qual o preço do carro em {Ano_do_carro2} '))
Ano_do_carro3 = Ano_do_carro2 + 1
preço_do_carro_ano3 = int(input(f'Qual o preço do carro em {Ano_do_carro3} '))

if preço_do_carro_ano1 > preço_do_carro_ano2 and preço_do_carro_ano1 > preço_do_carro_ano3:
  maior_valor = preço_do_carro_ano1
elif preço_do_carro_ano2 > preço_do_carro_ano3:
  maior_valor = preço_do_carro_ano2
else:
  maior_valor = preço_do_carro_ano3


if preço_do_carro_ano1 < preço_do_carro_ano2 and preço_do_carro_ano1 < preço_do_carro_ano3:
  menor_valor = preço_do_carro_ano1
elif preço_do_carro_ano2 < preço_do_carro_ano3:
  menor_valor = preço_do_carro_ano2
else:
  menor_valor = preço_do_carro_ano3

print(f'O maior valor do carro foi de : {maior_valor}  e o menor valor foi de {menor_valor}')


Qual o ano do carro? 2000
Qual o preço do carro em 2000 100
Qual o preço do carro em 2001 300
Qual o preço do carro em 2002 200
O maior valor do carro foi de : 300  e o menor valor foi de 100


5) Escreva um programa que pergunte sobre o preço de três produtos e indique qual é o produto mais barato para comprar.

In [ ]:
Produto1 = float(input('Qual o preço do produto? '))
Produto2 = float(input('Qual o preço do segundo produto? '))
Produto3 = float(input('Qual o preço do terceiro produto? '))

if Produto1 < Produto2 and Produto2 < Produto3:
  print( f'O preço mais barato é {Produto1}')
elif Produto2 < Produto3 :
  print( f'O preço mais barato é {Produto2}')
else:
  print( f'O preço mais barato é {Produto3}')

Qual o preço do produto? 30
Qual o preço do segundo produto? 10
Qual o preço do terceiro produto? 20
O preço mais barato é 10.0


6) Escreva um programa que leia três números e os exiba em ordem decrescente.

In [11]:
Num1 = int(input('Insira um número de 0 a 9: '))
Num2 = int(input('Insira um segundo número de 0 a 9: '))
Num3 = int(input('Insira um terceiro número de 0 a 9: '))

if Num1 > Num2 and Num1 > Num3:
  maior_valor = Num1
elif Num2 > Num1 and Num2 > Num3:
  maior_valor = Num2
else:
  maior_valor = Num3

if (Num1 < Num2 and Num1 > Num3) or (Num1 > Num2 and Num1 < Num3):
  valor_medio = Num1
elif (Num2 < Num1 and Num2 > Num3) or (Num2 > Num1 and Num2 < Num3):
  valor_medio = Num2
else:
  valor_medio = Num3

if Num1 < Num2 and Num1 < Num3:
  menor_valor = Num1
elif Num2 < Num1 and Num2 < Num3:
  menor_valor = Num2
else:
  menor_valor = Num3

print(maior_valor, valor_medio, menor_valor)

Insira um número de 0 a 9: 3
Insira um segundo número de 0 a 9: 2
Insira um terceiro número de 0 a 9: 1
3 2 1


7) Escreva um programa que pergunte em qual turno a pessoa usuária estuda ("manhã", "tarde" ou "noite") e exiba a mensagem "Bom Dia!", "Boa Tarde!", "Boa Noite!", ou "Valor Inválido!", conforme o caso.

In [22]:
periodo = input('Em qual turno você estuda? ').lower()
if periodo in ['manhã', 'dia' ,'de dia' , 'de manhã']:
  print('Bom dia!')
elif periodo in ['tarde', 'de tarde','a tarde'] :
  print('Boa tarde!')
elif periodo in ['noite', 'de noite', 'a noite']:
  print('Boa noite!')
else:
  print('Resposta inválida')


Em qual turno você estuda? Noite
Boa noite!


8) Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo %.

In [25]:
numero = int(input("Digite um número: "))

if numero % 2 == 0:
    print('Este numero é par.')
else:
    print('Este numero é par.')


Digite um número: 600
Este numero é par.


9) Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

In [50]:
numero = (input("Digite um número: "))
if '.'in numero  or ',' in numero:
  print('Este número é decimal')
else:
  print('Este número é inteiro')

Digite um número: 1.2
Este número é decimal


## Momento dos Projetos

10) Um programa deve ser escrito para ler dois números e, em seguida, perguntar à pessoa usuária qual operação ele deseja realizar. O resultado da operação deve incluir informações sobre o número - se é par ou ímpar, positivo ou negativo e inteiro ou decimal.

11) Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:

Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
Triângulo Equilátero: três lados iguais;
Triângulo Isósceles: quaisquer dois lados iguais;
Triângulo Escaleno: três lados diferentes.

12) Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R$ 2,00 e o preço do litro de etanol é R$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:

O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.

13) Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:

Para variação acima de 20%: bonificação para o time de vendas.
Para variação entre 2% e 20%: pequena bonificação para time de vendas.
Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
Para bonificações abaixo de -10%: corte de gastos.

# Estruturas de repetição

Quando temos que executar um mesmo bloco de comandos por várias vezes não é muito interessante fazer isso à mão.

Imaginemos a situação de termos que coletar e imprimir a média de duas notas de **3 estudantes**:

In [51]:
nota_1 = float(input('Digite a 1° nota: '))
nota_2 = float(input('Digite a 2° nota: '))

print(f'Média: {(nota_1+nota_2)/2}')
nota_1 = float(input('Digite a 1° nota: '))
nota_2 = float(input('Digite a 2° nota: '))

print(f'Média: {(nota_1+nota_2)/2}')
nota_1 = float(input('Digite a 1° nota: '))
nota_2 = float(input('Digite a 2° nota: '))

print(f'Média: {(nota_1+nota_2)/2}')

Digite a 1° nota: 6
Digite a 2° nota: 6
Média: 6.0
Digite a 1° nota: 10
Digite a 2° nota: 8
Média: 9.0
Digite a 1° nota: 4
Digite a 2° nota: 10
Média: 7.0


Agora imaginemos uma situação em que não são apenas 3 estudantes, mas sim 100 estudantes. Não seria interessante repetir o mesmo código por 100 vezes, mas sim **executar o mesmo código 100 vezes**.

Essa repetição conseguimos construir com laços de repetição!

## `WHILE`

O laço `while` é uma estrutura de controle de repetição em Python que permite executar um bloco de código repetidamente enquanto uma determinada condição é verdadeira. Sua estrutura é:



```
while condição:
    # bloco de código
```

Vamos construir um exemplo com um contador de 1 até 10.


In [52]:
contador = 1
while contador <= 10:
  print(contador)
  contador += 1

1
2
3
4
5
6
7
8
9
10


Agora vamos coletar as notas e médias de cada aluno dentro do while. Faremos um exemplo com 3 médias.

In [ ]:
contador = 1
while contador <= 3:
  nota_1 = float(input('Digite a 1° nota: '))
  nota_2 = float(input('Digite a 2° nota: '))

  print(f'Média: {(nota_1+nota_2)/2}')
  contador += 1

Digite a 1° nota: 5
Digite a 2° nota: 6
Média: 5.5
Digite a 1° nota: 7
Digite a 2° nota: 9
Média: 8.0
Digite a 1° nota: 4
Digite a 2° nota: 4
Média: 4.0


## `FOR`

O laço `for` é um tipo de estrutura de controle de fluxo em Python que permite iterar sobre um conjunto de elementos. A sua estrutura é:



```
for elemento in conjunto:
    # código a ser executado para cada elemento
```

O laço for itera sobre cada elemento do conjunto especificado e executa o bloco de código dentro do laço para cada elemento. Quando o laço chega ao final do conjunto, ele é interrompido e o programa continua a execução após o laço.



O conjunto pode ser gerado com a função [`range()`](https://docs.python.org/3/library/functions.html#func-range). Que é uma função capaz de gerar uma sequência de números inteiros. A estrutura dessa função é:

```
range(inicio, fim, passo)
```

Segundo a documentação, o `range()` gera uma sequência de números inteiros a partir do valor do parâmetro `inicio` até o valor do parâmetro `fim`, de acordo com o valor do parâmetro `passo`. Se `inicio` não for especificado, o valor padrão é 0. Se `passo` não for especificado, o valor padrão é 1.




Vamos fazer o mesmo contador `while` agora com `for`.

In [ ]:
for contador in range(1,11):
  print(contador)

1
2
3
4
5
6
7
8
9
10


In [ ]:
for contador in range(1,4):
  nota_1 = float(input('Digite a 1° nota: '))
  nota_2 = float(input('Digite a 2° nota: '))

  print(f'Média: {(nota_1+nota_2)/2}')

Digite a 1° nota: 4
Digite a 2° nota: 5
Média: 4.5
Digite a 1° nota: 6
Digite a 2° nota: 7
Média: 6.5
Digite a 1° nota: 5
Digite a 2° nota: 6
Média: 5.5


## Aquecendo na programação

1) Escreva um programa que peça dois números inteiros e imprima todos os números inteiros entre eles.



2) Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a B com 10.



3) Para tratar uma quantidade de 15 dados de avaliações de pessoas usuárias de um serviço da empresa, precisamos verificar se as notas são válidas. Então, escreva um programa que vai receber a nota de 0 a 5 de todos os dados e verificar se é um valor válido. Caso seja inserido uma nota acima de 5 ou abaixo de 0, repita até que a pessoa usuária insira um valor válido.



4) Desenvolva um programa que leia um conjunto indeterminado de temperaturas em Celsius e informe a média delas. A leitura deve ser encerrada ao ser enviado o valor -273°C.



5) Escreva um programa que calcule o fatorial de um número inteiro fornecido pela pessoa usuária. Lembrando que o fatorial de um número inteiro é a multiplicação desse número por todos os seus antecessores até o número 1. Por exemplo, o fatorial de 5 é 5 x 4 x 3 x 2 x 1 = 120.



# Momento dos projetos

6) Escreva um programa que gere a tabuada de um número inteiro de 1 a 10, de acordo com a escolha da pessoa usuária. Como exemplo, para o número 2, a tabuada deve ser mostrada no seguinte formato:

Tabuada do 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
COPIAR CÓDIGO



7) Os números primos possuem várias aplicações dentro da Ciência de Dados em criptografia e segurança, por exemplo. Um número primo é aquele que é divisível apenas por um e por ele mesmo. Assim, faça um programa que peça um número inteiro e determine se ele é ou não um número primo.




8) Vamos entender a distribuição de idades de pensionistas de uma empresa de previdência. Escreva um programa que leia as idades de uma quantidade não informada de clientes e mostre a distribuição em intervalos de [0-25], [26-50], [51-75] e [76-100]. Encerre a entrada de dados com um número negativo.




9) Em uma eleição para gerência em uma empresa com 20 pessoas colaboradoras, existem quatro candidatos(as). Escreva um programa que calcule o(a) vencedor(a) da eleição. A votação ocorreu da seguinte maneira:


Cada colaborador(a) votou em uma das quatro pessoas candidatas (que representamos pelos números 1, 2, 3 e 4).
Também foram contabilizados os votos nulos (representados pelo número 5) e os votos em branco (representados pelo número 6).
 Ao final da votação, o programa deve exibir o número total de votos para cada candidato(a), os nulos e os votos em branco. Além disso, deve calcular e exibir a porcentagem de votos nulos em relação ao total de votos e a porcentagem de votos em branco em relação ao total de votos.

# Estruturas de dados

Um conjunto de elementos é uma coleção de itens, que são armazenados juntos de maneira organizada. Alguns exemplos de conjuntos de elementos em Python são listas, strings e dicionários.

## Listas

As listas podem armazenar uma coleção de itens em ordem. Eles são delimitados por colchetes `[]` e os itens são separados por vírgulas.

Elas também podem armazenar qualquer tipo de item, incluindo números, strings, objetos e outras listas. Elas também podem armazenar itens de tipos de dados diferentes juntos em uma única lista.

In [5]:
lista = ['Fabricio Daniel',9.5,9.0,8.0,True]
lista

['Fabricio Daniel', 9.5, 9.0, 8.0, True]

In [2]:
type(lista)

list

As listas são organizadas em Python porque **cada elemento da lista tem um índice que indica sua posição na lista**. Os índices começam em 0 e vão até o tamanho da lista menos 1.

Temos então 5 elementos com índices variando de 0 a 4, ordenadamente:

```
#             [0]           [1]   [2]   [3]    [4]
lista = ['Fabricio Daniel', 9.5 , 9.0 , 8.0 , True]
```

Em Python temos também os índices **negativos** que se iniciam no último elemento com o valor de `-1` e depois avancam no universo dos negativos até chegar no 1° elemeno:

```
#             [-5]         [-4]  [-3]  [-2]   [-1]
lista = ['Fabricio Daniel', 9.5 , 9.0 , 8.0 , True]
```

Conseguimos selecionar separadamente cada elemento através de seus respectivos índices. Colocando o nome da lista e em seguida o índice a ser selecionado.

In [ ]:
lista[0]

'Fabricio Daniel'

In [ ]:
lista[1]

9.5

In [ ]:
lista[-1]

True

Uma forma mais dinâmica de trabalhar item por item de uma lista é utilizando um laço for para leitura elemento a elemento.

In [6]:
for elemento in lista:
  print(elemento)

Fabricio Daniel
9.5
9.0
8.0
True


A nota `8.0` de Fabricio Daniel precisa ser ajustada pois ganhou 2 pontos em sua ultima nota por fazer um trabalho de turma. Então é necessária fazer uma troca no valor do índice `3` de `8.0` para `10.0`.

In [7]:
lista[3] = 10.0
lista

['Fabricio Daniel', 9.5, 9.0, 10.0, True]

conseguimos calcular a média do aluno a partir dos dados que temos

In [8]:
media = (lista[1] + lista[2] + lista[3])/3
media

9.5

## Manipulação de listas

As listas são muito úteis em Python porque permitem armazenar e acessar uma coleção de itens de maneira organizada e rápida. Elas também oferecem muitos métodos úteis para manipular os itens armazenados, como adicionar, remover, classificar e pesquisar elementos.

#### Quantidade de elementos

Usamos a função [`len()`](https://docs.python.org/3/library/functions.html#len) para descobrimos a quantidade de elementos de um conjunto.

In [9]:
len(lista)

5

#### Partição

A partição de listas por indexação em Python é uma técnica muito útil para selecionar um subconjunto de elementos de uma lista. Ela é feita usando a sintaxe `lista[inicio:fim]`, onde `inicio` é o índice do primeiro elemento a ser incluído na partição e `fim` é o índice do primeiro elemento a ser excluído da partição.

In [10]:
lista[1:4]

[9.5, 9.0, 10.0]

In [11]:
lista[1:3]

[9.5, 9.0]

In [12]:
lista[:3]

['Fabricio Daniel', 9.5, 9.0]

In [16]:
lista[3:]

[10.0, True, 10.0, 8.0, 9.0]

In [17]:
lista[:]

['Fabricio Daniel', 9.5, 9.0, 10.0, True, 10.0, 8.0, 9.0]

#### [`append()`](https://docs.python.org/3/tutorial/datastructures.html#:~:text=of%20list%20objects%3A-,list.append(x),-Add%20an%20item)

Adiciona um elemento ao final da lista.

In [18]:
lista.append(media)
lista

['Fabricio Daniel', 9.5, 9.0, 10.0, True, 10.0, 8.0, 9.0, 9.5]

#### [`extend()`](https://docs.python.org/3/tutorial/datastructures.html#:~:text=list.extend(iterable))

Adiciona vários elementos ao final da lista.

Adicionaremos as notas `[10.0,8.0,9.0]` na lista do Fabricio Daniel

In [19]:
lista.extend([10.0,8.0,9.0])
lista

['Fabricio Daniel', 9.5, 9.0, 10.0, True, 10.0, 8.0, 9.0, 9.5, 10.0, 8.0, 9.0]

*Isso não é possivel ser feito com o* `append`.

In [20]:
lista.append([10.0,8.0,9.0])
lista

['Fabricio Daniel',
 9.5,
 9.0,
 10.0,
 True,
 10.0,
 8.0,
 9.0,
 9.5,
 10.0,
 8.0,
 9.0,
 [10.0, 8.0, 9.0]]

#### [`remove()`](https://docs.python.org/3/tutorial/datastructures.html#:~:text=append(x).-,list.remove(x),-Remove%20the%20first)

Remove um elemento específico da lista.

In [21]:
lista.remove([10.0,8.0,9.0])
lista

['Fabricio Daniel', 9.5, 9.0, 10.0, True, 10.0, 8.0, 9.0, 9.5, 10.0, 8.0, 9.0]

## Dicionário

Os dicionários são um tipo de estrutura de dados que armazenam pares de *chave-valor*. Eles são delimitados por chaves `{}` e os pares *chave-valor* são separados por vírgulas.

```
dicionário = {chave: valor}
```

A **chave** é um elemento único que identifica um valor no dicionário, enquanto o **valor** é o item que é armazenado para a chave. As chaves e os valores podem ser de **qualquer tipo de dado**.

Os dicionários são úteis para armazenar e acessar dados de maneira organizada e rápida. Eles são um tipo de conjunto de elementos em Python, pois armazenam uma coleção de itens.

In [22]:
dicionario = {'chave_1':1,
              'chave_2':2}
dicionario

{'chave_1': 1, 'chave_2': 2}

---
**Situação:**

Vamos criar um conjunto de dados com informações de matricula de um estudante. Os dados são os seguintes:

- matricula: 2000168933
- dia de cadastro: 25
- mês de cadastro: 10
- turma: 2E

In [23]:
cadastro = {'matricula': 2000168933,
            'dia_cadastro': 25,
            'mes_cadastro': 10,
            'turma': '2E'}
cadastro

{'matricula': 2000168933,
 'dia_cadastro': 25,
 'mes_cadastro': 10,
 'turma': '2E'}

In [24]:
cadastro['matricula']

2000168933

In [25]:
cadastro['turma']

'2E'

É possível substituir os valores dentro de uma chave. Por exemplo, recebemos a informação que a turma do estudante que cadastramos foi trocada para `'2G'` e agora precisamos trocar o valor da chave `'turma'`.

In [26]:
cadastro['turma'] = '2G'
cadastro

{'matricula': 2000168933,
 'dia_cadastro': 25,
 'mes_cadastro': 10,
 'turma': '2G'}

Podemos também adicionar outros dados ao dicionário. Vamos adicionar a informação de modalidade de ensino, nosso estudante atuará inicialmente em modalidade EAD.

Então iremos definir uma chave chamada `'modalidade'` e o valor `'EAD'`.

In [27]:
cadastro['modalidade'] = 'EAD'
cadastro

{'matricula': 2000168933,
 'dia_cadastro': 25,
 'mes_cadastro': 10,
 'turma': '2G',
 'modalidade': 'EAD'}

## Aprofundando em dicionários

#### [`pop()`](https://python-reference.readthedocs.io/en/latest/docs/dict/pop.html)
Remove um item de um dicionário e o retorna.

In [28]:
cadastro.pop('turma')

'2G'

In [29]:
cadastro

{'matricula': 2000168933,
 'dia_cadastro': 25,
 'mes_cadastro': 10,
 'modalidade': 'EAD'}

#### [`items()`](https://python-reference.readthedocs.io/en/latest/docs/dict/items.html)
Retorna uma lista de pares chave-valor do dicionário.

In [ ]:
cadastro.items()

dict_items([('matricula', 2000168933), ('dia_cadastro', 25), ('mes_cadastro', 10), ('modalidade', 'EAD')])

#### [`keys()`](https://python-reference.readthedocs.io/en/latest/docs/dict/keys.html)
Retorna uma lista das chaves do dicionário.

In [30]:
cadastro.keys()

dict_keys(['matricula', 'dia_cadastro', 'mes_cadastro', 'modalidade'])

#### [`values()`](https://python-reference.readthedocs.io/en/latest/docs/dict/values.html)
Retorna uma lista dos valores do dicionário.

In [31]:
cadastro.values()

dict_values([2000168933, 25, 10, 'EAD'])

### Leitura de valores com `for`

In [32]:
for chaves in cadastro.keys():
  print(cadastro[chaves])

2000168933
25
10
EAD


In [33]:
for valores in cadastro.values():
  print(valores)

2000168933
25
10
EAD


In [34]:
for chaves, valores in cadastro.items():
  print(chaves, valores)

matricula 2000168933
dia_cadastro 25
mes_cadastro 10
modalidade EAD


#Aquecendo na programação

1) Faça um programa que tenha a seguinte lista contendo os valores de gastos de uma empresa de papel

```
# [2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08].
```

 Com esses valores, faça um programa que calcule a média de gastos.

 Dica: use as funções built-in sum() e len().

2) Com os mesmos dados da questão anterior, defina quantas compras foram realizadas acima de 3000 reais e calcule a porcentagem quanto ao total de compras.

3) Faça um código que colete em uma lista 5 números inteiros quaisquer e imprima a lista. Exemplo: [1,4,7,2,4].

4) Colete novamente 5 inteiros e imprima a lista em ordem inversa à enviada.

5) Faça um programa que, ao inserir um número qualquer, cria uma lista contendo todos os números primos entre 1 e o número digitado.

6) Escreva um programa que peça uma data informando o dia, mês e ano e determine se ela é válida para uma análise.

# Momento dos projetos

7) Para um estudo envolvendo o nível de multiplicação de bactérias em uma colônia, foi coletado o número de bactérias por dia (em milhares) e pode ser observado a seguir: `[1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9].`

Tendo esses valores, faça um código que gere uma lista contendo o percentual de crescimento de bactérias por dia, comparando o número de bactérias em cada dia com o número de bactérias do dia anterior.

Dica: para calcular o percentual de crescimento usamos a seguinte equação:` 100 * (amostra_atual - amostra_passada) / (amostra_passada).`

8) Para uma seleção de produtos alimentícios, precisamos separar o conjunto de IDs dados por números inteiros sabendo que os produtos com ID par são doces e os com ID ímpar são amargos. Monte um código que colete 10 IDs. Depois, calcule e mostre a quantidade de produtos doces e amargos.

9) Desenvolva um programa que informa a nota de um(a) aluno(a) de acordo com suas respostas. Ele deve pedir a resposta desse(a) aluno(a) para cada questão e é preciso verificar se a resposta foi igual ao gabarito. Cada questão vale um ponto e existem as alternativas A, B, C ou D.

Gabarito da prova:
01 - D
02 - A
03 - C
04 - B
05 - A
06 - D
07 - C
08 - C
09 - A
10 - B

COPIAR CÓDIGO

10) Um instituto de meteorologia deseja fazer um estudo de temperatura média de cada mês do ano. Para isso, você precisa fazer um código que colete e armazene essas temperaturas médias em uma lista. Depois, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual e em que mês elas ocorreram, mostrando os meses por extenso (Janeiro, Fevereiro, etc.).

11) Uma empresa de e-commerce está interessada em analisar as vendas dos seus produtos. Os dados das vendas foram armazenados em um dicionário:

{'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
COPIAR CÓDIGO
Escreva um código que calcule o total de vendas e o produto mais vendido.

12) Uma pesquisa de mercado foi feita para decidir qual design de marca infantil mais agrada as crianças. A pesquisa foi feita e o votos computados podem ser observados abaixo:

'''
Tabela de votos da marca

Design 1 - 1334 votos

Design 2 - 982 votos

Design 3 - 1751 votos

Design 4 - 210 votos

Design 5 - 1811 votos

'''
COPIAR CÓDIGO

Adapte os dados fornecidos para uma estrutura de dicionário. A partir dele, informe o design vencedor e a porcentagem de votos recebidos.

13) As pessoas colaboradoras de um setor da empresa que você trabalha vão receber um abono correspondente a 10% do salário devido ao ótimo desempenho do time. O setor financeiro solicitou sua ajuda para a verificação das consequências financeiras que esse abono irá gerar nos recursos. Assim, foi encaminhada para você uma lista com os salários que receberão o abono: `[1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]`.

O abono de cada colaborador(a) não pode ser inferior a 200.
Em código, transforme cada um dos salários em chaves de um dicionário e o abono de cada salário no elemento.
Depois, informe o total de gastos com o abono, quantos(as) colaboradores(as) receberam o abono mínimo e qual o maior valor de abono fornecido.

14) Uma equipe de cientistas de dados está estudando a diversidade biológica em uma floresta. A equipe fez a coleta de informações sobre o número de espécies de plantas e animais em cada área dessa floresta e armazenou essas informações em um dicionário. Nele, a chave descreve a área dos dados e os valores nas listas correspondem às espécies de plantas e animais nas áreas, respectivamente.
```
{'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}

```
COPIAR CÓDIGO

Escreva um código para calcular a média de espécies por área e identificar a área com a maior diversidade biológica. Dica: use as funções built-in sum() e len().

15) O setor de RH da sua empresa te pediu uma ajuda para analisar as idades de colaboradores(as) de 4 setores da empresa. Para isso, foram fornecidos os seguintes dados:

```
{'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
 'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
 'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
 'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}

```

COPIAR CÓDIGO


Sabendo que cada setor tem 10 colaboradores(as), construa um código que calcule a média de idade de cada setor, a idade média geral entre todos os setores e quantas pessoas estão acima da idade média geral.